code/__HELPERS/AStar.dm

```diff
@@ -61,108 +64,129 @@ Actual Adjacent procs :
 	return b.f - a.f
 
 //wrapper that returns an empty list if A* failed to find a path
-/proc/get_path_to(caller, end, dist, maxnodes, maxnodedepth = 30, mintargetdist, adjacent = /turf/proc/reachableAdjacentTurfs, id=null, turf/exclude=null, simulated_only = 1)
+/proc/get_path_to(caller, end, dist, maxnodes, maxnodedepth = 30, mintargetdist, adjacent = /turf/proc/reachableTurftest, id=null, turf/exclude=null, simulated_only = 1)
 	var/list/path = AStar(caller, end, dist, maxnodes, maxnodedepth, mintargetdist, adjacent,id, exclude, simulated_only)
 	if(!path)
 		path = list()
 	return path
 
 //the actual algorithm
-/proc/AStar(caller, end, dist, maxnodes, maxnodedepth = 30, mintargetdist, adjacent = /turf/proc/reachableAdjacentTurfs, id=null, turf/exclude=null, simulated_only = 1)
-	var/list/pnodelist = list()
+/proc/AStar(caller, var/turf/end, dist, maxnodes, maxnodedepth = 30, mintargetdist, adjacent = /turf/proc/reachableAdjacentTurfs, id=null, turf/exclude=null, simulated_only = 1)
 	//sanitation
-	var/start = get_turf(caller)
+	var/turf/start = get_turf(caller)
 	if(!start)
 		return 0
-
+	if(start.z != end.z) //no PF between z levels
+		return 0
 	if(maxnodes)
 		//if start turf is farther than maxnodes from end turf, no need to do anything
 		if(call(start, dist)(end) > maxnodes)
 			return 0
 		maxnodedepth = maxnodes //no need to consider path longer than maxnodes
-
+	var/l
+	l = SSpathfinder.getfree()
+	while(!l)
+		stoplag(10)
```

Comment:
it is stoplag to check queue of pathfinding.if queue is free, this stoplag is ignored.

Reply:
Yes, but most pathfinds will not take 500ms so waiting 500ms when the path find currently being ran will likely only take 100ms to finish seems silly.

Reply:
Also, because how the elongating works if it wakes up after 500ms and the server is too lagging to return, it will actually sleep *again* for **double**, meaning that in times of lag, this will wait for 1500ms or 1.5 seconds.

This is why I recommend this be set to 2 or 3 ticks, 100 to 150ms of sleep time is more reasonable and won't create needless delays in path find processing during times of mild to moderate load.